thavalon-server/src/game/mod.rs

```diff
@@ -0,0 +1,247 @@
+        } else {
+            self.proposal_order[index + 1]
+        }
+    }
+
+    pub fn name(&self, player: usize) -> &str {
+        self.players[player].name.as_ref()
+    }
+
+    /// The number of players in the game
+    pub fn size(&self) -> usize {
+        self.proposal_order.len()
+    }
+}
+
+/// Fixed information about a player, decided at startup
+#[derive(Debug, Clone)]
+pub struct Player {
+    pub id: PlayerId,
+    pub name: String,
+    pub role: Role,
+}
+
+/// A collection of players, indexed in various useful ways.
+#[derive(Debug, Clone)]
+pub struct Players {
+    players: HashMap<PlayerId, Player>,
+    roles: HashMap<Role, PlayerId>,
+    good_players: Vec<PlayerId>,
+    evil_players: Vec<PlayerId>,
+}
+
+impl Players {
+    fn new() -> Players {
+        Players {
+            players: HashMap::new(),
+            roles: HashMap::new(),
+            good_players: Vec::new(),
+            evil_players: Vec::new(),
+        }
+    }
+
+    fn add_player(&mut self, player: Player) {
+        self.roles.insert(player.role, player.id);
+        if player.role.is_good() {
+            self.good_players.push(player.id);
+        } else {
+            self.evil_players.push(player.id);
+        }
+        self.players.insert(player.id, player);
+    }
+
+    fn by_role(&self, role: Role) -> Option<&Player> {
+        self.roles.get(&role).map(|id| &self[*id])
+    }
+
+    fn has_role(&self, role: Role) -> bool {
+        self.roles.contains_key(&role)
+    }
+
+    fn good_players(&self) -> &[PlayerId] {
+        self.good_players.as_slice()
+    }
+
+    fn evil_players(&self) -> &[PlayerId] {
+        self.evil_players.as_slice()
+    }
+
+    fn iter(&self) -> impl Iterator<Item = &Player> {
+        self.players.values()
+    }
+
+    fn len(&self) -> usize {
+        self.players.len()
+    }
+}
+
+impl Index<PlayerId> for Players {
+    type Output = Player;
+
+    fn index(&self, player_id: usize) -> &Self::Output {
+        &self.players[&player_id]
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum Card {
+    Success,
+    Fail,
+    Reverse,
+}
+
+impl fmt::Display for Card {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(match self {
+            Card::Success => "Success",
+            Card::Fail => "Fail",
+            Card::Reverse => "Reverse",
+        })
+    }
+}
+
+/// Game rules determined by the number of players
+#[derive(Debug, Clone)]
+pub struct GameSpec {
+    /// Number of players in the game
+    pub players: u8,
+    /// The number of players on each mission
+    pub mission_sizes: [usize; 5],
+    /// Allowed good roles in the game
+    pub good_roles: &'static [Role],
+    /// Allowed evil roles in the game
+    pub evil_roles: &'static [Role],
+    /// The number of players on the good team
+    pub good_players: u8,
+    /// True if mission 4 requires at least two failures
+    double_fail_mission_four: bool,
+}
+
+impl GameSpec {
```

Comment:
How easy would it be to have different rulesets. Could GameSpec be a trait that could be shared between different game rules?

Reply:
It probably depends on how different the rulesets are. We could probably have different `run_game` functions for different rulesets (ex. if proposal order is very different), and then reuse methods like `send_mission` and `get_proposal` among them, or move more info/logic into `GameSpec`